### ISSUE TITLE
Feature request: reduce unnecessary network traffic

### ISSUE THREAD

**Issue by kollokollo** [eclipse/mosquitto#1535]
I am referring to a case where N brokers build a network mirroring each others topics. 

1. It seems to me that a variant of QOS0 on topics with retained data should have an option
"publish at most once, but not if payload data is identical to (already stored) retained data." Maybe implemented with an additional flag. 

This would cause a temperature sensor reading the same temperature every second not causing all brokers to repeatly broadcast this again and again. Of course there might be cases where the time and the occurance itself of the identical data is important, but in other cases it is not.

2. To bridge, an additional mode should be configured where broker A can mirror a topic of broker B, but only actually subscribes to the topic on broker B, if at least on client is subscribed for it at broker A. And this should be handled dynamically. Maybe the bridging should be done always this way to save network bandwith. 

Currently broker A can subscribe to topics on broker B or push it own topics to it. In this case data is transmitted to broker B even if nobody listens there. 

Can you follow? This would be important when you set up a network of brokers all mirroring each other all topics on a slow/limited network connection. I hope I made it clear.

**Comment by gdt** [comment]
This is really a change request to the MQTT protocol and I think therefore out of scope for implementation in mosquitto.  Therefore i think this should be closed.

In addition a fresh temperature report with the same value is actually data.  (It seems like it would be best to address the data frequency and triggered updates at the sender, by slowing down the general rate and sending updates when the value changes.)

**Comment by kollokollo** [comment]
Yes, #1 would probably be a protocol change/extension. But you are assuming that there would be only a single producer(-device) publishing on the topic.  

#2 however, is an implementation issue.

**Comment by gdt** [comment]
True, but unrelated requests should be in separate issues, and then it would be less likely for someone to not blur them :-)

**Comment by karlp** [comment]
Item 2 would also miss retained messages.  The idea of "just save bandwidth" is nice, but it's not nearly as simple as you might think.

**Comment by kollokollo** [comment]
Hm, I cannot follow: How would retained messages be missed? When the mirroring broker resubscribes to the topic, it will get the retained message as a first one and can pass this to his subscribers (which also just that moment have subscribed).

The flow would be: 
0. Broker B has no subscribers to the topic and therefore unsubscribes from the topic on broker A, thus throw away any retained messge (or flag them  as old).
1. a client subscribes on broker B to the topic. There is no retained message at that moment. 
3. Immedeately broker B (re)subscribes at broker A to the topic and eventually receives a reteined message.
4. broker B forwards that retained message  to the cient (with retained flag set).

I do not see the problem.

Its never as simple. :-)

**Comment by ralight** [comment]
I'm closing this as out of scope
